src/libopensc/card-piv.c

```diff
@@ -2647,6 +2639,61 @@ static int piv_process_discovery(sc_card_t *card)
 	LOG_FUNC_RETURN(card->ctx, r);
 }
 
+
+/* normal way to get the discovery object via cache */
+static int piv_process_discovery(sc_card_t *card)
+{
+	int r;
+	u8 * rbuf = NULL;
+	size_t rbuflen = 0;
+
+	r = piv_get_cached_data(card, PIV_OBJ_DISCOVERY, &rbuf, &rbuflen);
+	/* Note rbuf and rbuflen are now pointers into cache */
+	if (r < 0)
+		goto err;
+
+	sc_log(card->ctx, "Discovery = %p:%"SC_FORMAT_LEN_SIZE_T"u", rbuf,
+	       rbuflen);
+	/* the object is now cached, see what we have */
+	r = piv_parse_discovery(card, rbuf, rbuflen, 0);
+
+err:
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+
+/* Do not use the cache value but read every time */
+static int piv_find_discovery(sc_card_t *card)
+{
+	int r = 0;
+	u8  rbuf[256];
+	size_t rbuflen = sizeof(rbuf);
+	u8 * arbuf = rbuf;
+	piv_private_data_t * priv = PIV_DATA(card);
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	/*
+	 * During piv_match or piv_card_reader_lock_obtained,
+	 * we use the discovery object to test if card present, and 
+	 * if PIV AID is active. So we can not use the cache
+	 */
+
+	/* If not valid, read, cache and test */
+	if (!(priv->obj_cache[PIV_OBJ_DISCOVERY].flags & PIV_OBJ_CACHE_VALID)) {
+		r = piv_process_discovery(card);
```

Comment:
The comment above states
> `/* Do not use the cache value but read every time */`
But here you're reading the cached value, right?

Reply:
No there is a "!" to say if the cache in not valid, call the piv_process_discovery. This routine will test if the cache is valid or not (in this case it will not be valid)  then read from the card add it to the cache and parse it find the AID. So the first time the discovery object is needed it will also get cached. 

If the cache was valid, then  the "else" in line 2685   will force the discovery object to be read from the card. piv_parse_discovery will then make sure the AID in the data read is the PIV AID, thus proving that the PIV application is active.

Reply:
The comment above the function let me believe that the function should "`not use the cache value`", but in your reply (and in your code) you're accessing the code. I'm just saying that there's an inconsistency. Maybe you just need to change the comment in the code.

Reply:
The `piv_process_discovery(card);` when called will return a cached value if it is cached. If not it will read the discovery object from the card and also addit it to the cache.

So the line `if (!(priv->obj_cache[PIV_OBJ_DISCOVERY].flags & PIV_OBJ_CACHE_VALID))`
firsts test if it is NOT cached, then calls  `piv_process_discovery(card);`
So in effect the discovery object will be read and also added to the cache. 

The code is working as intended.

Reply:
Could you then please remove or change the misleading comment *`/* Do not use the cache value but read every time */`*?

Reply:
Well ... the code and comment makes sense to me. The note, that we are reading the value always is already in the previous comment (`[...] So we can not use the cache`). This comment is supposed to explain just the two following lines, which it does.

The thing is that if **the cache is invalid**, we call `piv_process_discovery()` and it calls `piv_get_cached_data()`, which in the result **reads** the data from card again regardless the name of the function (we have invalid cache data or the object was not read yet).

I agree that the comment or the code can be little bit more clear.

But can not we explicitly invalidate the cache and then reduce this block to the call to `piv_process_discovery()` anyway (that is what we do in the else branch also, isn't it?). Would not it be more readable?

```
	/* invalidate the cache and read the object from card */
	priv->obj_cache[PIV_OBJ_DISCOVERY].flags &= ~PIV_OBJ_CACHE_VALID;
	r = piv_process_discovery(card); 
```
(I did not test this code, it might have some side effects or issues, but looks much more understandable and we even get some more debug information)